bitsail-connectors/connector-ftp/src/main/java/com/bytedance/bitsail/connector/ftp/source/reader/FtpSourceReader.java

```diff
@@ -0,0 +1,190 @@
+import com.bytedance.bitsail.connector.ftp.core.client.FtpHandlerFactory;
+import com.bytedance.bitsail.connector.ftp.core.client.IFtpHandler;
+import com.bytedance.bitsail.connector.ftp.core.config.FtpConfig;
+import com.bytedance.bitsail.connector.ftp.error.FtpErrorCode;
+import com.bytedance.bitsail.connector.ftp.source.split.FtpSourceSplit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+
+public class FtpSourceReader implements SourceReader<Row, FtpSourceSplit> {
+  private static final Logger LOG = LoggerFactory.getLogger(FtpSourceReader.class);
+
+  private final int subTaskId;
+
+  private final IFtpHandler ftpHandler;
+  private transient BufferedReader br;
+  private int totalSplitNum = 0;
+  private boolean hasNoMoreSplits = false;
+
+  private final Deque<FtpSourceSplit> splits;
+  private final transient DeserializationSchema<byte[], Row> deserializationSchema;
+
+  private FtpSourceSplit currentSplit;
+  private long currentReadCount;
+  private transient String line;
+
+  private FtpConfig ftpConfig;
+
+  private boolean skipFirstLine = false;
+  private String successFilePath;
+  private final transient Context context;
+
+  public FtpSourceReader(BitSailConfiguration jobConf, Context context, int subTaskId) {
+    this.subTaskId = subTaskId;
+    this.ftpConfig = new FtpConfig(jobConf);
+    this.ftpHandler = FtpHandlerFactory.createFtpHandler(ftpConfig);
+    this.splits = new LinkedList<>();
+    this.context = context;
+    this.deserializationSchema = DeserializationSchemaFactory.createDeserializationSchema(jobConf, ftpConfig, context);
+    LOG.info("FtpReader is initialized.");
+
+  }
+
+  @Override
+  public void start() {
+    this.ftpHandler.loginFtpServer();
+    checkPathsExist();
+    if (this.ftpConfig.getEnableSuccessFileCheck()) {
+      this.successFilePath = this.ftpConfig.getSuccessFilePath();
+      checkSuccessFileExist();
+    }
+    if (this.ftpHandler.getFtpConfig().getSkipFirstLine()) {
+      this.skipFirstLine = true;
+    }
+  }
+
+  @Override
+  public void pollNext(SourcePipeline<Row> pipeline) throws Exception {
+    if (line == null && splits.isEmpty()) {
+      return;
+    }
+    if (line == null) {
+      this.currentSplit = splits.poll();
+      LOG.info("Task {} begins to read split: {}={}", subTaskId, currentSplit.uniqSplitId(), currentSplit.getPath());
+      this.currentReadCount = 0;
+      ftpHandler.loginFtpServer();
+      InputStream in = ftpHandler.getInputStream(this.currentSplit.getPath());
+      br = new BufferedReader(new InputStreamReader(in, "utf-8"));
+      if (this.skipFirstLine) {
+        String skipLine = br.readLine();
+        LOG.info("Skip line:{}", skipLine);
+      }
+      line = br.readLine();
+    }
+    while (line != null) {
```

Comment:
looks like we read the whole file when calling the pollNext one time, this is a bit not compatible with the current reader model. For example, we will see reading a single record taking too long on the metric. Can we just pull one record in each `pollNext` call?

Reply:
You are right, I will update it in the next commit.